Add `TimestampWithOffset` canonical extension type

serramatutu opened the pull request:
# Which issue does this PR close?

Implements the new `arrow.timestamp_with_offset` canonical extension type for arrow-go.

# Rationale for this change

Be compatible with the Arrow columnar spec's `arrow.timestamp_with_offset` canonical extension type.

# What changes are included in this PR?

Adds a `TimestampWithOffset` extension type. This type represents a timestamp column that stores a potentially different timezone offset per value: the timestamp is stored in UTC alongside the original timezone offset in minutes. The offset-in-minutes field can be primitive-, dictionary-, or run-end-encoded.

**This PR has been decoupled from the inner-nullability parity work it was previously based on.** Those comparison / JSON-marshaling changes now live in #918 and can be reviewed and merged independently. This PR is now scoped to just the extension type and touches only:

- `arrow/extensions/timestamp_with_offset.go` (new)
- `arrow/extensions/timestamp_with_offset_test.go` (new)
- `arrow/extensions/extensions.go` (register the canonical type)

To keep this PR self-contained, the JSON round-trip test asserts the values the type guarantees (per-row instant, timezone offset, and validity) rather than `array.RecordEqual`, which would compare inner non-nullable struct child validity bitmaps — the parity concern tracked separately in #918.

# Are these changes tested?

Yes — primitive/dictionary/run-end offset encodings, JSON marshal + round-trip, and IPC round-trip.

# Are there any user-facing changes?

Yes, this is a new canonical extension type.

serramatutu committed:
Add `TimestampWithOffset` extension type

This commit adds a new `TimestampWithOffset` extension type that can be
used to represent timestamps with per-row timezone information. It
stores information in a `struct` with 2 fields, `timestamp=[T, "UTC"]`,
where `T` can be any `arrow.TimeUnit` and `offset_minutes=int16`, which
represents the offset in minutes from the UTC timestamp.

serramatutu committed:
Allow dictionary encoding of `TimestampWithOffset`

This commit allows `TimestampWithOffset` to be dict-encoded.

 - I made `NewTimestampWithOffsetType` take in an input
  `offsetType arrow.DataType`. It returns an error if the data type is not
  valid.

- I added a new infallible `NewTimestampWithOffsetTypePrimitiveEncoded`
  to make the encoding explicit.

- I added `NewTimestampWithOffsetTypeDictionaryEncoded` which returns an
  error in case the given type is not a valid dictionary key type.

- I made all tests run in a for loop with all possible allowed encoding
  types, ensuring all encodings work.

serramatutu committed:
Allow run-end encoding of `TimestampWithOffset`

serramatutu committed:
Optimize `Values()` and `MarshalJSON()` for REE

Smartly iterate over offsets if they're run-end encoded instead of doing
a binary search at every iteration.

This makes the loops O(n) instead of O(n*logn).

serramatutu committed:
General code smell patches

Changed a lot of things based on Matt's suggestions.

serramatutu committed:
Test with `-08:30` and `11:00` timezone

serramatutu commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,568 @@
+		}
+	})
+	return values
+}
```
Fixed: https://github.com/apache/arrow-go/pull/558/changes/bd12287760927080eda309594b36dc9e2ec813ff

serramatutu committed:
Rename constructors

serramatutu committed:
Leverage generics for REE and dict-encoding

serramatutu committed:
Make equals also check for data type

serramatutu commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,568 @@
+	}
+
+	v, _ := NewTimestampWithOffsetType(timeUnit, offsetType)
+	// SAFETY: the offsetType has already been checked by isDataTypeCompatible, so we can ignore the error
+
+	return v, nil
+}
+
+func (b *TimestampWithOffsetType) ExtensionEquals(other arrow.ExtensionType) bool {
+	return b.ExtensionName() == other.ExtensionName()
+}
```
Fixed: https://github.com/apache/arrow-go/pull/558/changes/2762d64b50221ac82ed0a22231eb8478427823ed

serramatutu replied:
Did this in a separate PR: https://github.com/apache/arrow-go/pull/758

zeroshade commented on arrow/array/encoded.go:
```diff
@@ -386,6 +386,15 @@ func (b *RunEndEncodedBuilder) ContinueRun(n uint64) {
+func (b *RunEndEncodedBuilder) UnsafeAppendBoolToBitmap(v bool) {
+	if !v {
+		b.finishRun()
+	}
```
this seems incorrect, it could be a run of nulls, couldn't it? And wouldn't the run itself be handled by the value append? I don't think we need to call `b.finishRun` here

zeroshade commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,577 @@
+	case *arrow.Int16Type:
+		return true
+	case *arrow.DictionaryType:
+		return arrow.IsInteger(offsetType.IndexType.ID()) && arrow.TypeEqual(offsetType.ValueType, arrow.PrimitiveTypes.Int16)
```
`arrow.IsInteger` is unnecessary here, by definition only integer types are allowed for the IndexTypes of a dictionary type. Anything else is invalid anyways, so there's no need to check that.

serramatutu replied:
Fixed: https://github.com/apache/arrow-go/pull/558/changes/f42417b775645a0fd4b4ec2502acb1a327cc9222

zeroshade commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,568 @@
+	// naive "bitwise" conversion to UTC, keeping the underlying date the same
+	utc := t.UTC()
+	naiveUtc := time.Date(t.Year(), t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), time.UTC)
+	offsetMinutes := int16(naiveUtc.Sub(t).Minutes())
```
can you give me an example where it messes up the values? 

Just to verify, the goal here is to take `t` and provide the same *instant of time* in UTC while figuring out what the offset was in minutes for the timezone, yes? i.e. given `2009-11-10 15:00:00 -0800 PST` this should return `2009-11-10 23:00:00 +0000 UTC` and `480`, correct?

serramatutu replied:
Actually I'm a dumdum. Fixed: https://github.com/apache/arrow-go/pull/558/changes/3a9a8f2af3bb37058418a2ebb94ae0bdd4e7ed5e

zeroshade commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,568 @@
+	switch {
+	case a.IsNull(i):
+		return array.NullValueStr
+	default:
+		return a.Value(i).String()
+	}
+}
+
+func (a *TimestampWithOffsetArray) MarshalJSON() ([]byte, error) {
+	values := make([]interface{}, a.Len())
+	a.iterValues(func(i int, timestamp *time.Time) {
+		values[i] = timestamp
+	})
+	return json.Marshal(values)
```
Using `0` as a sentinel for `null` seems incorrect as that means you can't actually encode a non-null value of 1970-01-01 and get the correct output. Since we need to be able to represent the difference between a *value* of `0` and a `null`, then either `iterValues()` needs to return a sequence of `*time.Time` or it needs to be a sequence of `iter.Seq2[time.Time, bool]` where the bool indicates validity (if the bool is false, then it's null).

serramatutu replied:
https://github.com/apache/arrow-go/pull/558/changes/d955cdb46fe2d17c547e51f9f5e123b2edcb7d76

Fix here.

zeroshade commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,577 @@
+			}
+		} else {
+			for i := 0; i < a.Len(); i++ {
+				ts := time.Unix(0, 0)
```
this should be equivalent to just doing `var ts time.Time`, except faster since it avoids the extra processing of calling the `Unix` function.

zeroshade commented on arrow/extensions/timestamp_with_offset_test.go:
```diff
@@ -0,0 +1,358 @@
+	// assert.False(t, extensions.NewTimestampWithOffsetType(arrow.Microsecond).ExtensionEquals(extensions.NewTimestampWithOffsetTypeRunEndEncoded(arrow.Microsecond, &arrow.Int16Type{})))
+	// assert.False(t, extensions.NewTimestampWithOffsetTypeDictionaryEncoded(arrow.Microsecond, &arrow.Int16Type{}).ExtensionEquals(extensions.NewTimestampWithOffsetTypeRunEndEncoded(arrow.Microsecond, &arrow.Int16Type{})))
+	//
+	// // Dict-encoding key type is not equal
```
why are these all commented out?

serramatutu replied:
I was dumb. Fixed https://github.com/apache/arrow-go/pull/558/changes/02498067c12dfd04265234e715f8c8da34d22fbc

serramatutu committed:
Remove outdated docs

serramatutu committed:
Remove unnecessary IsInteger call

serramatutu committed:
Remove specific type for `TimestampWithOffsetRunEndsType`

serramatutu committed:
Uncomment test

serramatutu commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,577 @@
+// Struct(timestamp=Timestamp(T, "UTC"), offset_minutes=Dictionary(I, Int16)), where T is any TimeUnit and I is a
+// valid Dictionary index type.
+//
+// The error will be populated if the index is not a valid dictionary-encoding index type.
```
Fixed: https://github.com/apache/arrow-go/pull/558/changes/76e3ace8d45e9162706282203eff81766db8dd97

serramatutu commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,577 @@
+// Struct(timestamp=Timestamp(T, "UTC"), offset_minutes=RunEndEncoded(E, Int16)), where T is any TimeUnit and E is a
+// valid run-ends type.
+//
+// The error will be populated if runEnds is not a valid run-end encoding run-ends type.
```
Fixed: https://github.com/apache/arrow-go/pull/558/changes/76e3ace8d45e9162706282203eff81766db8dd97

serramatutu commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,577 @@
+
+type TimestampWithOffsetRunEndsType interface {
+	*arrow.Int8Type | *arrow.Int16Type | *arrow.Int32Type | *arrow.Int64Type |
+		*arrow.Uint8Type | *arrow.Uint16Type | *arrow.Uint32Type | *arrow.Uint64Type
```
Fixed: https://github.com/apache/arrow-go/pull/558/changes/231cb1ec4211cf65d8e96fa3c5beec1a94bfc940

serramatutu replied:
Actually I had to revert. Spec says those types are different:

Dict index is either signed or unsigned int of 8-64 bits
> The index type of a Dictionary type can only be an integer type, preferably signed, with width 8 to 64 bits.

Run ends is a signed int of 16-64 bits
> The run end type of a Run-End Encoded type can only be a signed integer type with width 16 to 64 bits.

https://arrow.apache.org/docs/format/Columnar.html#data-types

serramatutu committed:
Fix REE ends type

serramatutu committed:
Support unix epoch

Stop using `time=0` as a sentinel for null in `iterValues()`. Instead,
return a tuple of `(time.Time, bool)` where the bool indicates whether the
value is valid or not.

Added a test for it.

serramatutu committed:
Fix inconsistency with `UnsafeAppend()`

serramatutu committed:
Fix lint

serramatutu committed:
Stop appending nulls to child arrays

serramatutu commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,568 @@
+				offsets.Append(offsetMinutes)
+			} else {
+				timestamps.UnsafeAppendBoolToBitmap(false)
+				offsets.UnsafeAppendBoolToBitmap(false)
```
I did change this but now the IPC roundtrip tests are failing because the internal count of nulls is different between the thing before IPC and after IPC.

To make them pass we'll need to relax the `arrow.RecordEqual()` to ignore `NullN()` when the field is not nullable. Alternatively we can change it so that an array builder sets `nulls = 0` if the array is not nullable, regardless of what has been set before in calls to `UnsafeAppendBoolToBitmap()` etc

Related: https://lists.apache.org/thread/7gbqjwykh1ob3xbvwph3ljsdl5c7kxpd

serramatutu committed:
Simplify `fieldValuesFromTime()` implementation

serramatutu committed:
Add subtests

zeroshade committed:
Fix run-end-encoded offset builder in TimestampWithOffset AppendValues

Start a run when none exists yet (e.g. leading null rows) instead of
continuing a nonexistent run, and only update lastOffset when a new run
starts. This stops a leading null from corrupting the run-ends/values
children and stops a null between equal offsets from splitting one
contiguous run into two. Add a noLastOffset sentinel, handle nil valids,
and add regression tests across all offset encodings.

zeroshade committed:
Fix TimestampWithOffset builder nil validity and reuse handling

Address roborev review findings on the run-end-encoded offset builder:

- AppendValues now treats a nil valids slice as all-valid so the parent
  struct and its child builders stay the same length. Previously the
  struct received zero slots while the children received len(values),
  producing an inconsistent extension array.
- Override NewArray/NewExtensionArray to reset lastOffset to noLastOffset
  on finalization, so a reused builder starts a fresh run instead of
  continuing a run that belonged to the array just finalized (which could
  emit run-ends without a matching value on a run-end-encoded offset).
- Add regression tests for nil validity and builder reuse across all
  offset encodings.

zeroshade committed:
Treat empty validity slice as all-valid in TimestampWithOffset builder

Follow-up to the nil-validity fix: an empty (len 0) valids slice is also
all-valid, matching the convention of the other Arrow builders.
AppendValues now normalizes both nil and empty validity, and panics on a
genuine length mismatch (a non-empty valids whose length differs from
values).

Extend the validity regression test to cover the empty slice and add a
test asserting the length-mismatch panic.

zeroshade committed:
Reset REE run tracker on TimestampWithOffset builder null paths

Follow-up to the run-end-encoding fixes: AppendNull (and therefore the
UnmarshalOne(null) and AppendValueFromString(null) paths that call it)
appends a null offset run via the embedded struct builder but left
lastOffset stale. A subsequent value repeating the pre-null offset would
then continue the null run instead of starting its own, encoding that
value with the null offset run.

Override AppendNull/AppendNulls to reset lastOffset to noLastOffset, and
add a regression test for Append(value), AppendNull(), Append(same-offset
value) with run-end-encoded offsets.

zeroshade commented on arrow/array/compare.go:
```diff
@@ -35,9 +39,17 @@ func RecordEqual(left, right arrow.RecordBatch) bool {
 
 	for i := range left.Columns() {
+		lf := left.Schema().Field(i)
+		rf := left.Schema().Field(i)
```
Addressed in the current head: `recordEqual` now takes `rf := right.Schema().Field(i)` and gates on `schemaFieldEqual(lf, rf)` (name + nullability), so left/right field differences are detected. Resolving.

zeroshade commented on arrow/array/compare.go:
```diff
@@ -47,16 +59,26 @@ func RecordEqual(left, right arrow.RecordBatch) bool {
-
 	for i := range left.Columns() {
+		lf := left.Schema().Field(i)
+		rf := left.Schema().Field(i)
```
Addressed: `recordApproxEqual` now reads `rf := right.Schema().Field(i)` and checks `schemaFieldEqual(lf, rf)` before comparing, so nullability/name differences are caught. Resolving.

zeroshade commented on arrow/array/compare.go:
```diff
@@ -160,21 +194,29 @@ func TableEqual(left, right arrow.Table) bool {
-		rc := right.Column(i)
-		if !lc.Field().Equal(rc.Field()) {
+		lf := left.Schema().Field(i)
+		rf := left.Schema().Field(i)
```
Addressed: `tableEqual` now uses `rf := right.Schema().Field(i)` with a `schemaFieldEqual(lf, rf)` guard. Resolving.

zeroshade commented on arrow/array/compare.go:
```diff
@@ -183,27 +225,35 @@ func TableApproxEqual(left, right arrow.Table, opts ...EqualOption) bool {
-		rc := right.Column(i)
-		if !lc.Field().Equal(rc.Field()) {
+		lf := left.Schema().Field(i)
+		rf := left.Schema().Field(i)
```
Addressed: `tableApproxEqual` now uses `rf := right.Schema().Field(i)` with the same `schemaFieldEqual(lf, rf)` guard. Resolving.

zeroshade commented on arrow/array/record_test.go:
```diff
@@ -511,14 +516,20 @@ func TestRecordBuilder(t *testing.T) {
 		}
 	}
 
+	err := b.UnmarshalJSON([]byte(`{"f1-i32": 6, "f2-f64-notnull": 6.6, "map": [{"key": "4": "value": "d"}]}`))
```
Fixed: the literal now reads `{"key": "4", "value": "d"}` (valid JSON) instead of the malformed colon, so the test exercises the intended path. Resolving.

zeroshade commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,535 @@
+			} else {
+				offsets.ContinueRun(1)
+			}
+			b.lastOffset = offsetMinutes16
```
Fixed in the builder rework. Both `Append` and `AppendValues` now start a run with `offsets.Append(1)` whenever none exists yet (`b.lastOffset == noLastOffset`), so `ContinueRun(1)` is never called without a matching value entry — including leading-null rows. `lastOffset` is only updated on the branch that starts a new run, so a null row can't split a run or shift a boundary based on a masked value, and `AppendNull`/`AppendNulls`/`NewArray` reset the tracker. There's now an inline comment documenting exactly this case. Resolving.

serramatutu reviewed (commented):
@zeroshade some comments:

Can you think of a way to decouple these 2 things (timestamp with offset and the inner nullability fixes)? So we can merge the extension type without having to wait on all the fixes for that and do that more carefully?

I'm sorry I left the branch based off of an old version of the inner nullability parity fixes. There are some differences with what I was doing in the new version of the [JSON PR](https://github.com/apache/arrow-go/pull/833):
- I ended up dropping the `equalOpts` because this is not what C++ does. It looks like they just validate the schemas on `Validate()` instead of making `Equal()` more lenient. I had not fixed that in that PR because there are some issues with the CSV reader (it calls `Validate()`) and it was becoming somewhat of a sprawl to fix it all. There are lots of behavior in `arrow-go` that never checks for the schema of structs.
- I did not do the `GetOneForMarshalNullable()` and instead I changed the struct/record builder to check for `IsValid()` before calling `GetOneForMarshal()`.

zeroshade committed:
Decouple JSON roundtrip test from inner-nullability comparison

Replace array.RecordEqual with a per-row check of the extension's
guarantees (instant, timezone offset, validity). RecordEqual compares the
inner non-nullable struct child validity bitmaps, which legitimately differ
between builder output (children valid under a null parent) and JSON reader
output (children null). That inner-nullability parity is tracked separately
and is out of scope for this extension type.

zeroshade commented:
@serramatutu updated this PR to be *only* the timestamp with offset extension type

zeroshade committed:
Fix run-end-encoded offset decoding for sliced TimestampWithOffset arrays

iterValues started the run-end tracker at physical index 0 and compared
the logical row against absolute run ends, ignoring the array offset. A
sliced array whose offset began inside a later run decoded offsets from
the wrong run in Values() and MarshalJSON(). Initialize the tracker from
GetPhysicalOffset() and advance using absolute positions (Offset()+i),
and add a regression test slicing an REE-encoded array across run
boundaries.

zeroshade commented:
@serramatutu filed #918 to split out the nullability stuff

zeroshade committed:
Address review comments on TimestampWithOffset

- timeFromFieldValues: derive the zone-name sign from the whole offset so
  a sub-hour negative offset keeps its sign (e.g. -30 minutes now formats
  as "UTC-00:30" instead of "UTC+00:30"; integer hours is 0 there and
  cannot carry the sign). Add a regression test.
- MarshalJSON: store the timestamp by value instead of taking the address
  of the range variable.
- Values: rename the loop variable so it no longer shadows the time
  package.

zeroshade commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,592 @@
+		default:
+			fmt.Fprintf(&o, "\"%s\"", a.Value(i))
+		}
+	}
+	o.WriteString("]")
+	return o.String()
+}
+
+func timeFromFieldValues(utcTimestamp arrow.Timestamp, offsetMinutes int16, unit arrow.TimeUnit) time.Time {
+	hours := offsetMinutes / 60
+	minutes := offsetMinutes % 60
+	if minutes < 0 {
+		minutes = -minutes
+	}
+
+	loc := time.FixedZone(fmt.Sprintf("UTC%+03d:%02d", hours, minutes), int(offsetMinutes)*60)
+	return utcTimestamp.ToTime(unit).In(loc)
+}
```
Fixed in e7ef6d8d: the zone-name sign is now derived from the whole offset, so a sub-hour negative offset keeps its sign (e.g. -30 minutes now formats as "UTC-00:30" instead of "UTC+00:30"). Added TestTimestampWithOffsetSubHourNegativeZone as a regression test.

zeroshade commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,592 @@
+	}
+}
+
+func (a *TimestampWithOffsetArray) MarshalJSON() ([]byte, error) {
+	values := make([]interface{}, a.Len())
+	i := 0
+	for ts, valid := range a.iterValues() {
+		if !valid {
+			values[i] = nil
+		} else {
+			values[i] = &ts
+		}
+		i += 1
+	}
+	return json.Marshal(values)
+}
```
Addressed in e7ef6d8d: MarshalJSON now stores the timestamp by value instead of the address of the range variable. (For the record, under this module's Go 1.25 per-iteration loop semantics it was not actually an aliasing bug — the existing exact-JSON roundtrip test asserts distinct per-row values — but marshaling by value is clearer.)

zeroshade commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,592 @@
+	}
+}
+
+func (a *TimestampWithOffsetArray) Values() []time.Time {
+	return slices.Collect(func(yield func(time.Time) bool) {
+		for time := range a.iterValues() {
+			if !yield(time) {
+				return
+			}
+		}
+	})
+}
```
Fixed in e7ef6d8d: renamed the loop variable to `t` so it no longer shadows the `time` package.

zeroshade committed:
Avoid int16 overflow when formatting TimestampWithOffset zone name

Widen offsetMinutes to int before negating so an out-of-range int16
offset (e.g. math.MinInt16) cannot overflow when computing the absolute
value used for the zone name.

serramatutu commented on arrow/extensions/timestamp_with_offset_test.go:
```diff
@@ -0,0 +1,591 @@
+			require.NoError(t, err)
+			defer roundtripped.Release()
+			require.Equal(t, schema, roundtripped.Schema())
+			// Avoid array.RecordEqual here: it compares the inner non-nullable
```
Can we add a FIXME here to revisit? Ideally the records SHOULD equal after roundtripping lol

zeroshade replied:
done

zeroshade committed:
fix(extensions): reject non-empty metadata in TimestampWithOffset Deserialize

Deserialize now rejects serialized metadata other than "" or "{}", matching the canonical extension spec (empty metadata) and the existing JSONType pattern. Adds a regression test.

Also documents (NOTE #918) that AppendNull writes a null into the non-nullable offset_minutes storage field; inner-field comparison/JSON parity is tracked in #918.

zeroshade committed:
test(extensions): add negative conformance tests for TimestampWithOffset Deserialize

Adds TestTimestampWithOffsetTypeDeserializeInvalidStorage covering non-struct storage, wrong field count/names/order, non-UTC and empty timezone, nullable timestamp/offset fields, and invalid offset encodings (wrong primitive type, dictionary value not Int16, run-end-encoded value not Int16).

zeroshade committed:
test(extensions): add FIXME to revisit RecordEqual on TimestampWithOffset roundtrip

lidavidm commented on arrow/extensions/timestamp_with_offset.go:
```diff
@@ -0,0 +1,608 @@
+	case *arrow.RunEndEncodedType:
+		return offsetType.ValidRunEndsType(offsetType.RunEnds()) &&
+			arrow.TypeEqual(offsetType.Encoded(), arrow.PrimitiveTypes.Int16)
+		// FIXME: Technically this should be non-nullable, but a Arrow IPC does not deserialize
+		// ValueNullable properly, so enforcing this here would always fail when reading from an IPC
+		// stream
+		// !offsetType.ValueNullable
```
This is pretty major right? Can we link this to a tracking issue? (Does integration catch this?)

zeroshade replied:
This is part of what was pulled out into #918 and friends
